### PR TITLE
SCALE + ASH methods

### PR DESCRIPTION
**Pull request by paulnovello**
Implements both ASH and SCALE baselines at the same time because they are very similar.

# SCALE method

**SCALE** method basically consists of re-using existing logit-based OOD methods, but with penultimate layer **activations scaled**. Let $a$ be the activation vector, and $P_p(a)$ the $p$-th percentile of $a$'s values. The scaling is computed using the formula
$s = \exp(\frac{\sum_{i} a_i}{\sum_{a_i > P_p(a)} a_i})$

Here, we focus on a Resnet trained on CIFAR10, challenged on SVHN.

**Reference**  
_Scaling for Training Time and Post-hoc Out-of-distribution Detection Enhancement_, ICLR 2024
<https://arxiv.org/abs/2111.12797>

# ASH method

This notebook aims at evaluating the **ASH method**.

**ASH** method basically consists of re-using existing logit-based OOD methods, but with penultimate layer **activations scaled and pruned**. Let $a$ be the activation vector, and $P_p(a)$ the $p$-th percentile of $a$'s values. The scaling is computed using the same formula as above. The activation vector is pruned for values $a_i \leq P_p(a)$.

Here, we focus on a Resnet trained on CIFAR10, challenged on SVHN.

**Reference**  
_Extremely Simple Activation Shaping for Out-of-Distribution Detection_, ICLR 2023
<http://arxiv.org/abs/2209.09858>

### CODE REVIEW

File: oodeel/extractor/torch_feature_extractor.py [hunk 1]
```diff
@@ -317,6 +335,50 @@ def hook(_, __, output):
 
         return hook
 
+    def _get_scale_hook(self, percentile: float) -> Callable:
+        """
+        Hook that truncate activation features under a threshold value
```
Comment (cofri) [inline oodeel/extractor/torch_feature_extractor.py:340]:
```suggestion
        Hook that truncates activation features under a threshold value
```

File: oodeel/extractor/torch_feature_extractor.py [hunk 2]
```diff
@@ -317,6 +335,50 @@ def hook(_, __, output):
...
+        Returns:
+            Callable: hook function
+        """
+
+        def hook(_, __, output):
+            output_percentile = torch.quantile(output, percentile, dim=1)
+            mask = output > output_percentile[:, None]
+            output_masked = output * mask
+            s = torch.exp(torch.sum(output, dim=1) / torch.sum(output_masked, dim=1))
+            s = torch.unsqueeze(s, 1)
```
Comment (cofri) [inline oodeel/extractor/torch_feature_extractor.py:354]: Same suggestion, avoid duplicate code between both hooks?

File: tests/tests_tensorflow/methods/test_tf_ash.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    "detector_name,auroc_thr,fpr95_thr",
+    [
+        ("odin", 0.5, 0.5),
+        ("mls", 0.5, 0.5),
+        ("energy", 0.5, 0.5),
+    ],
+)
+def test_ash(detector_name, auroc_thr, fpr95_thr):
+    """
+    Test ASH + [MLS, MSP, Energy, ODIN, Entropy] on toy blobs OOD dataset-wise task
```
Comment (cofri) [inline tests/tests_tensorflow/methods/test_tf_ash.py:41]: Why don't you test Entropy and MSP, as suggested in the docstring here? BTW, we could now add GEN.

File: tests/tests_tensorflow/methods/test_tf_ash.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+from oodeel.methods import ODIN
+from tests.tests_tensorflow import eval_detector_on_blobs
+
+
+@pytest.mark.parametrize(
+    "detector_name,auroc_thr,fpr95_thr",
+    [
+        ("odin", 0.5, 0.5),
+        ("mls", 0.5, 0.5),
+        ("energy", 0.5, 0.5),
```
Comment (cofri) [inline tests/tests_tensorflow/methods/test_tf_ash.py:36]: 0.5 as AUROC and FPR95 seems a very loose threshold. Is there a reason?

File: tests/tests_torch/methods/test_torch_ash.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+from oodeel.methods import ODIN
+from tests.tests_torch import eval_detector_on_blobs
+
+
+@pytest.mark.parametrize(
+    "detector_name,auroc_thr,fpr95_thr",
+    [
+        ("odin", 0.5, 0.5),
+        ("mls", 0.5, 0.5),
+        ("energy", 0.5, 0.5),
```
Comment (cofri) [inline tests/tests_torch/methods/test_torch_ash.py:36]: Same comment as for TensorFlow test

**Review by cofri — CHANGES_REQUESTED** [review]
Nice big PR!

File: tests/tests_tensorflow/methods/test_tf_ash.py (hunk 3 above)
Comment (paulnovello) [reply tests/tests_tensorflow/methods/test_tf_ash.py:41]: Because Entropy and MSP relies on softmax and seems to induce numerical instabilities and Nans. It is the same for GEN. Maybe we should put in the doc that ASH should not be used with Entropy MSP and GEN.

File: tests/tests_torch/methods/test_torch_ash.py (hunk 5 above)
Comment (paulnovello) [reply tests/tests_torch/methods/test_torch_ash.py:36]: Same answer ;)

File: oodeel/extractor/torch_feature_extractor.py (hunk 2 above)
Comment (paulnovello) [reply oodeel/extractor/torch_feature_extractor.py:354]: I agreed for keras because the duplicate code was longer and within the same method, but in that case, it is not so long, and I think it would hurt readability

Comment (cofri) [reply oodeel/extractor/torch_feature_extractor.py:356]: Yes, I agree for torch. Let's keep it that way

File: tests/tests_tensorflow/methods/test_tf_ash.py (hunk 3 above)
Comment (cofri) [reply tests/tests_tensorflow/methods/test_tf_ash.py:41]: Ok, indeed it can be good to add a warning message somewhere to warn users about ASH

File: tests/tests_tensorflow/methods/test_tf_ash.py (hunk 4 above)
Comment (cofri) [reply tests/tests_tensorflow/methods/test_tf_ash.py:36]: Issue https://github.com/deel-ai/oodeel/issues/95 opened by @paulnovello.
Comment resolved
